### PR TITLE
Add support for bitwise inner-product in painless

### LINKED ISSUE #111232: Investigate & implement support for different vector types dot-product against bit vectors
### Description

Back when Cohere originally released their v3 model, one of the interesting steps they took for relevance was reranking the top-N vectors by doing a dot-product between the `float[]` and the `bit[]` vectors. Effectively summing all the non-zero indices, allowing for the higher-fidelity of the query vector to provide better scoring.

We should dig a little bit to see if this is worth it with our own benchmarking with CohereV3 and see if there are any other models would benefit from this.

The work likely required are:

 - Reproducing Cohere's recommendation: https://github.com/cohere-ai/BinaryVectorDB/blob/main/BinaryVectorDB/BinaryVectorDB.py
 - Scripting method definition changes to allow float & byte to be dot-product values to `bit` vector types
 - Panama Vector SIMD work so that the functions are not dog slow (using the bits as a bit-mask and then summing the lanes)

What's interesting is that we might want to allow dot-product between all our various element_type's in scripts. The tricky part is ensuring we always pick the right type.

### PR DESCRIPTION
**Pull request by benwtrent**
This adds bitwise inner product to painless. 

The idea here is:

 - For two bit arrays, which we determine to be a byte array whose dimensions match `dense_vector.dim/8`, we simply return bitwise `&`
 - For a stored bit array (remember, with `dense_vector.dim/8` bytes), sum up the provided byte or float array using the bit array as a mask.

This is effectively supporting asynchronous quantization. A prime example of how this works is: https://github.com/cohere-ai/BinaryVectorDB

Basically, you do your initial search against the binary space and then rerank with a differently quantized vector allowing for more information without additional storage space. 

closes:  https://github.com/elastic/elasticsearch/issues/111232

### CODE REVIEW

File: libs/simdvec/src/main/java/org/elasticsearch/simdvec/ESVectorUtil.java [hunk 1]
```diff
@@ -24,4 +34,95 @@ public static long ipByteBinByte(byte[] q, byte[] d) {
...
+
+    /** AND bit count striding over 8 bytes at a time. */
+    static int andBitCountLong(byte[] a, byte[] b) {
+        int distance = 0, i = 0;
+        for (final int upperBound = a.length & -Long.BYTES; i < upperBound; i += Long.BYTES) {
+            distance += Long.bitCount((long) BitUtil.VH_NATIVE_LONG.get(a, i) & (long) BitUtil.VH_NATIVE_LONG.get(b, i));
+        }
+        // tail:
+        for (; i < a.length; i++) {
+            distance += Integer.bitCount((a[i] & b[i]) & 0xFF);
```
Comment (rjernst) [inline libs/simdvec/src/main/java/org/elasticsearch/simdvec/ESVectorUtil.java:124]: Could the tail be done with a single `Long.bitCount` call, if using a mask based on the number of remaining bytes?

File: libs/simdvec/src/main/java/org/elasticsearch/simdvec/ESVectorUtil.java [hunk 2]
```diff
@@ -24,4 +34,95 @@ public static long ipByteBinByte(byte[] q, byte[] d) {
...
+        }
+        return distance;
+    }
+
+    /** AND bit count striding over 8 bytes at a time. */
+    static int andBitCountLong(byte[] a, byte[] b) {
+        int distance = 0, i = 0;
+        for (final int upperBound = a.length & -Long.BYTES; i < upperBound; i += Long.BYTES) {
```
Comment (rjernst) [inline libs/simdvec/src/main/java/org/elasticsearch/simdvec/ESVectorUtil.java:119]: > a.length & -Long.BYTES

I assume you are trying to calculate the number of full longs that can fit in `a.length`? This is a clever trick, but it deserves a comment explaining the goal and how it is achieved.

File: libs/simdvec/src/main/java/org/elasticsearch/simdvec/ESVectorUtil.java [hunk 3]
```diff
@@ -24,4 +34,95 @@ public static long ipByteBinByte(byte[] q, byte[] d) {
...
+     *
+     * @param a bytes containing a vector
+     * @param b bytes containing another vector, of the same dimension
+     * @return the value of the AND bit count of the two vectors
+     */
+    public static int andBitCount(byte[] a, byte[] b) {
+        if (a.length != b.length) {
+            throw new IllegalArgumentException("vector dimensions differ: " + a.length + "!=" + b.length);
+        }
+        if (XOR_BIT_COUNT_STRIDE_AS_INT) {
```
Comment (rjernst) [inline libs/simdvec/src/main/java/org/elasticsearch/simdvec/ESVectorUtil.java:96]: Since this is statically known, could which method to use be also chosen statically? It could be captured in a method handle and called here with invokeExact?

**Review by rjernst — COMMENTED** [review]
A couple thoughts

File: libs/simdvec/src/main/java/org/elasticsearch/simdvec/ESVectorUtil.java (hunk 1 above)
Comment (benwtrent) [reply libs/simdvec/src/main/java/org/elasticsearch/simdvec/ESVectorUtil.java:145]: Possibly? But I didn't want to bother with over optimizing. Especially since these methods are effectively copy-pastes of what exists in Lucene for xor (just changing to `&`).

File: libs/simdvec/src/main/java/org/elasticsearch/simdvec/ESVectorUtil.java [hunk 4]
```diff
@@ -24,4 +47,103 @@ public static long ipByteBinByte(byte[] q, byte[] d) {
...
+                    result += q[i * Byte.SIZE + j];
+                }
+            }
+        }
+        return result;
+    }
+
+    /**
+     * AND bit count computed over signed bytes.
+     * Copied from Lucene's XOR implementation
```
Comment (john-wagster) [inline libs/simdvec/src/main/java/org/elasticsearch/simdvec/ESVectorUtil.java:100]: This is more so for my education.  What's the thinking here for putting this in ES vs Lucene? given that we have XOR in Lucene VectorUtil and this seem complementary.

**Review by john-wagster — APPROVED** [review]
LGTM

File: libs/simdvec/src/main/java/org/elasticsearch/simdvec/ESVectorUtil.java (hunk 4 above)
Comment (benwtrent) [reply libs/simdvec/src/main/java/org/elasticsearch/simdvec/ESVectorUtil.java:100]: @john-wagster there is no compelling reason for keeping it out of Lucene. But, its weird for there to be public utility methods in Lucene when nothing directly utilizes it.

**Comment by benwtrent** [comment]
@elasticmachine update branch

File: docs/reference/vectors/vector-functions.asciidoc
Comment (mayya-sharipova) [inline docs/reference/vectors/vector-functions.asciidoc:1]: On line 19 we also say that `dot_product` is not supported for bit vectors.

File: docs/reference/vectors/vector-functions.asciidoc [hunk 5]
```diff
@@ -332,6 +332,9 @@ When using `bit` vectors, not all the vector functions are available. The suppor
 * <<vector-functions-hamming,`hamming`>> – calculates Hamming distance, the sum of the bitwise XOR of the two vectors
 * <<vector-functions-l1,`l1norm`>> – calculates L^1^ distance, this is simply the `hamming` distance
 * <<vector-functions-l2,`l2norm`>> - calculates L^2^ distance, this is the square root of the `hamming` distance
+* <<vector-functions-dot-product,`dotProduct`>> – calculates dot product. When comparing two `bit` vectors,
```
Comment (mayya-sharipova) [inline docs/reference/vectors/vector-functions.asciidoc:335]: May be we can add that queryVector can be byte[] (of the same dims as docs or dims *8), or also can be a string, and can be of float[]

**Review by mayya-sharipova — APPROVED** [review]
@benwtrent Thanks Ben, great change! I've added a small docs comment.

File: docs/reference/vectors/vector-functions.asciidoc (hunk 5 above)
Comment (benwtrent) [reply docs/reference/vectors/vector-functions.asciidoc:335]: ++

**Comment by benwtrent** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/script/VectorScoreScriptUtils.java [hunk 6]
```diff
@@ -307,6 +307,84 @@ public interface DotProductInterface {
...
+            float[] floatQueryVector = new float[queryVector.size()];
+            byte[] byteQueryVector = new byte[queryVector.size()];
+            boolean isFloat = false;
+            for (int i = 0; i < queryVector.size(); i++) {
+                Number number = queryVector.get(i);
+                floatQueryVector[i] = number.floatValue();
+                byteQueryVector[i] = number.byteValue();
+                if (floatQueryVector[i] % 1.0f != 0.0f || floatQueryVector[i] < Byte.MIN_VALUE || floatQueryVector[i] > Byte.MAX_VALUE) {
+                    isFloat = true;
+                }
```
Comment (tteofili) [inline server/src/main/java/org/elasticsearch/script/VectorScoreScriptUtils.java:349]: I think we can `break` here

**Review by tteofili — APPROVED** [review]
LGTM (with minor comment)

File: server/src/main/java/org/elasticsearch/script/VectorScoreScriptUtils.java (hunk 6 above)
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/script/VectorScoreScriptUtils.java:349]: We need to build both vectors and then pick the right now, breaking would prevent us from building the arrays.
